Server/src/services/dataProcess/JsonFileHandler.ts

```diff
@@ -0,0 +1,200 @@
+import { BadRequest } from '@tsed/exceptions';
+import { validationSchema } from '../helpers/validationSchema';
+import { IJsonDatasetModel } from '../../genericInterfaces/DataProcessInterfaces'
+const fileSystem = require('fs');
+
+
+export class JsonFileFactory extends FileHandlerFactory {
+
+    getFileHandler(filePath: string): AbstractFileHandler {
+        return new JsonFileHandler(filePath)
+    }
+}
+
+export class JsonFileHandler extends AbstractFileHandler {
+
+    async parseFile(): Promise<IJsonDatasetModel> {
+        try {
+            this.parsedFileData = await JSON.parse(fileSystem.readFileSync(this.filePath))
+            return this.parsedFileData
+        } catch (err) {
+            throw new BadRequest("Cannot parse your file. Something is wrong with it");
+        }
+    }
+
+    async validateExtractedData(jsonData: any): Promise<JSON> {
+        try {
+            await validationSchema.validate(jsonData)
+            return jsonData;
+        } catch (err) {
+            throw new BadRequest(err.message);
+        }
+    }
+
+    async uploadData(jsonData: any): Promise<string> {
+
+        let category: string = '';
+        let subcategory: string = '';
+        let dataSetName: string = '';
+        let dataType: string = '';
+        let dataSetComments: string = '';
+        let individualDataSetComments: string[] = [];
+        let material: IMaterials[] = [];
+        let referenceType: string = '';
+        let referencePublisher: string = '';
+        let referenceTitle: string = '';
+        let referenceDOI: string = '';
+        let referenceAuthors: IAuthors[] = [];
+        let referenceYear: number;
+        let referencePages: number;
+        let referenceVolume: number;
+        let referenceTypeID: number;
+        let referenceDatePublished: Date = null;
+        let referenceDateAccessed: Date = null;
+        let publisherNameId: number;
+        let publicationID: number;
+        let dataSetDataTypeID: number;
+        let datasetID: number;
+        let unitsID: number;
+        let reprID: number;
+
+        this.parsedFileData = jsonData
+
+        // Create this object after the parsing passes
+        let uploadModel = new DataUploadModel();
+
+        try {
+            referenceTypeID = await uploadModel.insertReferenceType(referenceType); console.log('Received reference ID' + referenceTypeID);
+        } catch (err) {
+            console.log('rejected request for referenceTypeID');
+        }
+
+        referencePublisher = this.parsedFileData.reference.publisher;
+        try {
+            publisherNameId = await uploadModel.insertPublisher(referencePublisher);
+            console.log('Received publisher name ID' + publisherNameId);
+        } catch (err) {
+            console.log('rejected request for inserting publisherNameId')
+        }
+
+
+        referenceAuthors = this.parsedFileData.reference.authors;
+        let allAuthors: Authors[];
+        try {
```

Comment:
i would prefer if all the fields that you are uploading would have their own functions, this function is very long, and hard to keep it all in mind

Reply:
A+